File: java/1904/Solution.java
class Solution {
    private static final int MAX_ROUND = 96;

    public int numberOfRounds(String loginTime, String logoutTime) {
        int loginTimestamp = this.parseTime(loginTime);
        int logoutTimestamp = this.parseTime(logoutTime);

        int firstRoundNumber = this.calculateFirstRound(loginTimestamp);
        int lastRoundNumber = this.calculateLastRound(logoutTimestamp);

        if (loginTimestamp > logoutTimestamp) {
            return MAX_ROUND + lastRoundNumber - firstRoundNumber;
        }

        if (lastRoundNumber - firstRoundNumber < 0) {
            return 0;
        }

        return lastRoundNumber - firstRoundNumber;
    }

    private int calculateFirstRound(int startTime) {
        if (startTime % 15 == 0) {
            return startTime / 15;
        }

        return (startTime / 15) + 1;
    }

    private int calculateLastRound(int endTime) {
        return endTime / 15;
    }

    private int parseTime(String time) {
        String[] timeToken = time.split(":");

        int hour = Integer.parseInt(timeToken[0]);
        int minute = Integer.parseInt(timeToken[1]);

        return (hour * 60) + minute;
    }
}

File: python3/1904/Solution.py
class Solution:
    def __init__(self):
        self.MAX_ROUND = 96

    def numberOfRounds(self, loginTime: str, logoutTime: str) -> int:
        login_timestamp = self.parseTime(loginTime)
        logout_timestamp = self.parseTime(logoutTime)

        first_round_number = self.calculateFirstRound(login_timestamp)
        last_round_number = self.calculateLastRound(logout_timestamp)

        if login_timestamp > logout_timestamp:
            return self.MAX_ROUND + last_round_number - first_round_number

        if last_round_number < first_round_number:
            return 0

        return last_round_number - first_round_number

    def calculateFirstRound(self, start_time: int) -> int:
        if start_time % 15 == 0:
            return start_time // 15

        return (start_time // 15) + 1

    def calculateLastRound(self, end_time: int) -> int:
        return end_time // 15

    def parseTime(self, time: str) -> int:
        time_token = time.split(':')

        hour = int(time_token[0])
        minute = int(time_token[1])

        return (hour * 60) + minute
